src/late_multicellular_stage/DualContourer.cs

```diff
@@ -0,0 +1,423 @@
+        var triIndices = new List<int>(capacity);
+
+        // Shape point is located at intersection of three lines of the grid
+        // if a point's value is true, the point is inside shape, otherwise outside
+        var shapePoints = new bool[gridTo.X - gridFrom.X + 2,
+            gridTo.Y - gridFrom.Y + 2,
+            gridTo.Z - gridFrom.Z + 2];
+
+        // Points at grid cell centers at coordinates of gridpoint are related to shape points with coordinates of:
+        // [gridpoint.X; gridpoint.X + 1] for x, [gridpoint.Y; gridpoint.Y + 1] for y,
+        // [gridpoint.Z; gridpoint.Z + 1] for z
+
+        sw.Stop();
+
+        // GD.Print($"Allocated memory in {sw.Elapsed}");
+        sw.Restart();
+
+        CalculatePoints(shapePoints, gridFrom, gridTo);
+
+        /*sw.Stop();
+        GD.Print($"Calculated points in {sw.Elapsed}");
+        sw.Restart();*/
+
+        Vector3I gridOffset = -gridFrom;
+
+        for (int x = gridFrom.X; x <= gridTo.X; ++x)
+        {
+            for (int y = gridFrom.Y; y <= gridTo.Y; ++y)
+            {
+                for (int z = gridFrom.Z; z <= gridTo.Z; ++z)
+                {
+                    // var realPos = new Vector3(x, y, z) / PointsPerUnit;
+                    var gridPos = new Vector3I(x, y, z);
+
+                    var tris = GetBestTriangles(x + gridOffset.X, y + gridOffset.Y, z + gridOffset.Z, shapePoints);
+
+                    if (tris == null)
+                        continue;
+
+                    PlaceTriangles(gridPos,
+                        tris,
+                        points, triIndices, placedPoints);
+                }
+            }
+        }
+
+        /*sw.Stop();
+        GD.Print($"Placed triangles in {sw.Elapsed}");
+        sw.Restart();*/
+
+        var normals = new Vector3[points.Count];
+
+        if (Smoothen)
+        {
+            AdjustVertices(points);
+            AdjustVertices(points, 0.25f, normals);
+        }
+
+        /*sw.Stop();
+        GD.Print($"Adjusted vertices in {sw.Elapsed}");*/
+
+        var colors = new Color[points.Count];
+
+        SetColours(points, colors);
+
+        var arrays = new Array();
+        arrays.Resize((int)Mesh.ArrayType.Max);
+
+        arrays[(int)Mesh.ArrayType.Vertex] = points.ToArray();
+        arrays[(int)Mesh.ArrayType.Index] = triIndices.ToArray();
+        arrays[(int)Mesh.ArrayType.Normal] = normals;
+        arrays[(int)Mesh.ArrayType.Color] = colors;
+
+        // arrays[(int)Mesh.ArrayType.TexUV] = newUV;
+        // arrays[(int)Mesh.ArrayType.TexUV2] = newUV1;
+
+        ArrayMesh mesh = new ArrayMesh();
+        mesh.AddSurfaceFromArrays(Mesh.PrimitiveType.Triangles, arrays);
+
+        return mesh;
+    }
+
+    private void CalculatePoints(bool[,,] shapePoints, Vector3I gridFrom, Vector3I gridTo)
+    {
+        var sw = new Stopwatch();
+        sw.Start();
+
+        Vector3I gridOffset = -gridFrom;
+
+        Parallel.For(gridFrom.X, gridTo.X + 1, delegate(int x)
```

Comment:
This seems also pretty questionable to be, and should really be unrolled into a fixed number of tasks to use with our TaskExecutor.

Reply:
Should be done now too